edalize/cocotb.py

```diff
@@ -0,0 +1,157 @@
+        simulator_options = self.tool_options.get('tool_options', {})
+        if self.option_name[simulator_name] in simulator_options:
+            simulator_options[self.option_name[simulator_name]] += [cocotb_args]
+        else:
+            simulator_options[self.option_name[simulator_name]] = [cocotb_args]
+
+        edam['tool_options'][simulator_name] = simulator_options
+
+        # Instantiate the required simulator backend
+        self.simulator = get_edatool(simulator_name)(edam, work_root)
+
+        # Run methods that rely on delegation to simulator backend
+        self._configure_environment()
+
+    @classmethod
+    def get_doc(cls, api_ver):
+        if api_ver == 0:
+            return {'description': 'cocotb - see <https://docs.cocotb.org/> for additional documentation on options',
+                    'members': [
+                        {'name': 'simulator',
+                         'type': 'String',
+                         'desc': 'The simulator backend to use to run Cocotb (required)'},
+                        {'name': 'module',
+                         'type': 'String',
+                         'desc': 'Comma-separated list of the Python modules to search for test functions (required)'},
+                        {'name': 'toplevel_lang',
+                         'type': 'String',
+                         'desc': 'Language of top level HDL module (required)'},
+                        {'name': 'random_seed',
+                         'type': 'Integer',
+                         'desc': 'Seed the Python random module to recreate a previous test stimulus'},
+                        {'name': 'testcase',
+                         'type': 'String',
+                         'desc': 'Comma-separated list the test functions to run'},
+                        {'name': 'cocotb_results_file',
+                         'type': 'String',
+                         'desc': 'The file name where xUnit XML tests results are stored'},
+                    ]
+                    }
+
+    # This creates a python path that contains the parent directory of every
+    # python file in the fileset. This is not ideal because several directories
+    # in a hierarchical tree may be added, which allows a given module to be
+    # imported in several ways. One way to fix this in future might be to allow
+    # path components to be passed into Edalize from FuseSoC as well as files.
+    def _create_python_path(self):
+        (src_files, incdirs) = self._get_fileset_files()
+        path_components = []
+
+        for f in src_files:
+            if f.file_type == 'pythonSource':
+                component = os.path.join(self.work_root, os.path.dirname(f.name))
+                if component not in path_components:
+                    path_components.append(component)
+
+        return os.pathsep.join(path_components)
+
+    def _configure_environment(self):
+        os.environ['MODULE'] = self.tool_options['module']
```

Comment:
Would be probably better to not set this globally but push into subprocess that executes.

Reply:
I agree, but I'm not sure I can easily do that here. The subprocess is executed by the simulator backend and I don't think there's a way to inject the environment changes into that. Unless I've missed something, which is possible!

One approach might be to add an execution environment to the edam config that gets passed to the simulator backend. That might require more extensive changes that I'd like to make without some discussion though.

Reply:
On second thoughts, this might be possible by redefining _run_tool() in my Cocotb class to set an environment.